pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -469,106 +443,160 @@ pub async fn disk_usage_eviction_task_iteration_impl<U: Usage>(
+            if !tl.is_active() {
+                continue;
+            }
+            let info = tl.get_local_layers_for_disk_usage_eviction();
+            debug!(timeline_id=%tl.timeline_id, "timeline resident layers count: {}", info.resident_layers.len());
+            tenant_candidates.extend(
+                info.resident_layers
+                    .into_iter()
+                    .map(|layer_infos| (tl.clone(), layer_infos)),
+            );
+            max_layer_size = max_layer_size.max(info.max_layer_size.unwrap_or(0));
 
-    let min_resident_size = match mode {
-        Mode::GlobalLru => {
-            lru_candidates.append(scratch);
-            return ControlFlow::Continue(());
+            if cancel.is_cancelled() {
+                return Ok(EvictionCandidates::Cancelled);
+            }
         }
-        Mode::RespectTenantMinResidentSize => tenant
-            .get_min_resident_size_override()
-            .unwrap_or(max_layer_size),
-    };
 
-    scratch.sort_unstable_by_key(|(_, layer_info)| layer_info.last_activity_ts);
+        // `min_resident_size` defaults to maximum layer file size of the tenant.
+        // This ensures that each tenant can have at least one layer resident at a given time,
+        // ensuring forward progress for a single Timeline::get in that tenant.
+        // It's a questionable heuristic since, usually, there are many Timeline::get
+        // requests going on for a tenant, and, at least in Neon prod, the median
+        // layer file size is much smaller than the compaction target size.
+        // We could be better here, e.g., sum of all L0 layers + most recent L1 layer.
+        // That's what's typically used by the various background loops.
+        //
+        // The default can be overriden with a fixed value in the tenant conf.
+        // A default override can be put in the default tenant conf in the pageserver.toml.
+        let min_resident_size = if let Some(s) = tenant.get_min_resident_size_override() {
+            debug!(
+                "using overridden min resident size {} for tenant {}",
+                s,
+                tenant.tenant_id()
+            );
+            s
+        } else {
+            debug!(
+                "using max layer size {} for tenant {}",
+                max_layer_size,
+                tenant.tenant_id()
+            );
+            max_layer_size
+        };
 
-    let mut current: u64 = scratch.iter().map(|(_, layer)| layer.file_size()).sum();
-    for (tl, layer) in scratch.drain(..) {
-        if cancel.is_cancelled() {
-            return ControlFlow::Break(());
-        }
-        if current <= min_resident_size {
-            break;
+        // Sort layers most-recently-used first, then partition by
+        // cumsum above/below min_resident_size.
+        tenant_candidates
+            .sort_unstable_by_key(|(_, layer_info)| std::cmp::Reverse(layer_info.last_activity_ts));
+        let mut cumsum: i128 = 0;
```

Comment:
Thinking we probably should have whatever other name for this.

Reply:
I think it's quite a common name for rolling sum.